### PR TITLE
genpolicy: hardening some agent requests

### PR DESCRIPTION
**Pull request by burgerdev**
This PR backports the commits mentioned in #10046, except for the symlink source (which I don't quite understand yet) and the sandbox storage checks (which are likely blocked by #8833). The commit messages each have a rationale for why this hardening makes sense.

cc @danmihai1

### CODE REVIEW

**Comment by danmihai1** [comment]
@burgerdev , the only problem I see with this PR is that it doesn't include tests to demonstrate that the attacks that should be mitigated are actually mitigated. But, if other reviewers are OK without these tests, then merging this PR is OK with me too. I agree these changes are useful - thanks!

(I was hoping to add tests by using https://github.com/orgs/kata-containers/projects/44 - but sadly we haven't been able to get that done yet.)

**Comment by danmihai1** [comment]
@burgerdev , if you get a chance please:

1. Try to understand the new test failures on TDX. Either @Redent0r or I will help if needed.

2. Rebase on top of the latest main branch code - because we have just enabled genpolicy testing on SEV and SNP. To recap, genpolicy is tested by CI nowadays on:
- cbl-mariner
- tdx
- sev
- snp

3. Watch out for possible test failures related to UpdateEphemeralMountsRequest. When I tried to ban that API in the main branch several months ago, Kata CI still needed UpdateEphemeralMountsRequest to be enabled. Probably because they were not always using shared_fs=none in Kata settings. I don't know if this is still a significant issue.

4. (minor) reorder CreateSandboxRequest to match https://github.com/microsoft/kata-containers/blob/386cab09d8c544097f47c60591b283cff2397439/src/tools/genpolicy/rules.rego#L1163. This should help my team a bit, when we'll pick up your changes downstream.

Thanks for all your help!

**Review by danmihai1 — CHANGES_REQUESTED** [review]
I noticed some UpdateEphemeralMountsRequest errors on cbl-mariner. We can try to sort these out here, or in https://github.com/kata-containers/kata-containers/pull/9911 - either way would be OK with me.

**Comment by burgerdev** [comment]
Thanks for the feedback!

1. Let's do `UpdateEphemeralMountsRequest` in #9911, I dropped it here.
2. Rebased on today's main.
3. Added an integration test for the `CopyFile` method.
   * The idea is to take canned requests (which are very simple for `CopyFile`) and evaluate the policy against them. If we agree on such tests being useful, I can add more for other methods.
   * This is not yet integrated into anything, pointers on where to do that would be much appreciated (in the `test` target in the Makefile probably - anywhere else?).
5. Acknowledged the diff between msft and main, will fix that too.

**Comment by danmihai1** [comment]
Thanks @burgerdev !

Looks like TDX got stuck at some CopyFile calls - please take a look.

The "make test" you suggested seems good at a quick look. Cc @Redent0r too. Introducing this new type of testing seems to deserve a separate PR.

When making changes to the CopyFile rules, adding some test cases in tests/integration/kubernetes/k8s-policy-*.bats seems useful too.

**Comment by burgerdev** [comment]
@Redent0r: I spotted the double-anchor problem with the regex substitution, but looking at the input paths I now wonder: why did we need to [modify `cpath` for TDX](https://github.com/kata-containers/kata-containers/blob/d69950e5c6eaa1b97a1e7d15de5a3229fa413857/tests/integration/kubernetes/tests_common.sh#L138 ) specifically? Isn't it [statically defined](https://github.com/kata-containers/kata-containers/blob/d69950e5c6eaa1b97a1e7d15de5a3229fa413857/src/agent/src/storage/fs_handler.rs#L38)?

@danmihai1 I can move the test to another PR, sure, but in what order should we then go about it?

Regarding the `tests/integration/kubernetes/k8s-policy-*.bats` tests: I missed that there are now single agent RPC probes, will add some for `CopyFileRequest`! However, I find it incredibly difficult to work with these - are there instructions somewhere for spinning up a dev environment where one can quickly iterate on these?

**Comment by danmihai1** [comment]
> @danmihai1 I can move the test to another PR, sure, but in what order should we then go about it?

I would add the "test infrastructure" first, and then add the CopyFile test content.

> 
> Regarding the `tests/integration/kubernetes/k8s-policy-*.bats` tests: I missed that there are now single agent RPC probes, will add some for `CopyFileRequest`! However, I find it incredibly difficult to work with these - are there instructions somewhere for spinning up a dev environment where one can quickly iterate on these?

To iterate on those tests, my team is using the steps we've learned from kata-containers/tests/integration/kubernetes/gha-run.sh. My example below works from my Azure Linux OS, but other distros should work very similarly:

```
az extension add --name aks-preview

az login --use-device-code

my_resource_group=<Some resource group name>

az group create --name "${my_resource_group}" --location eastus
```

Get DOCKER_TAG from https://github.com/kata-containers/kata-containers/actions/workflows/ci-nightly.yaml  -> kata-containers-ci-on-push / run-k8s-tests-on-aks / run-k8s-tests (cbl-mariner, clh, normal) -> get kata-tarball -> DOCKER_TAG. Example:

```
DOCKER_TAG=a7931115a017cb38ad805ad3031076b0198b9002-nightly-amd64
GH_PR_NUMBER=nightly
GOROOT=/usr/local/go && export GOPATH=$HOME/gopath && export PATH=$GOPATH/bin:$GOROOT/bin:$PATH

git clone https://github.com/kata-containers/kata-containers.git
cd kata-containers/tests/integration/kubernetes/

K8S_TEST_HOST_TYPE=normal \
AZ_RG="${my_resource_group}" \
DOCKER_REGISTRY=ghcr.io \
DOCKER_REPO=kata-containers/kata-deploy-ci \
DOCKER_TAG="${DOCKER_TAG}" \
GH_PR_NUMBER="${GH_PR_NUMBER}" \
KATA_HYPERVISOR=clh \
KBS_INGRESS=aks \
KUBERNETES=vanilla \
USING_NFD=false \
bash gha-run.sh create-cluster
```

Get your cluster name from your AKS GUI. Then:

```
my_cluster_name=<your cluster name>

az aks get-credentials --resource-group "${my_resource_group}" --name ${my_cluster_name} --overwrite-existing

AZ_RG="${my_resource_group}" \
DOCKER_REGISTRY=ghcr.io \
DOCKER_REPO=kata-containers/kata-deploy-ci \
DOCKER_TAG="${DOCKER_TAG}" \
GH_PR_NUMBER="${GH_PR_NUMBER}" \
KBS_INGRESS=aks \
KUBERNETES=vanilla \
USING_NFD=false \
KATA_HYPERVISOR=clh \
bash gha-run.sh deploy-kata-aks
```

Set-up genpolicy for CI testing:

```
pushd kata-containers/src/tools/genpolicy

cargo build && \
sudo mkdir -p /opt/kata/bin && \
sudo mkdir -p /opt/kata/share/defaults/kata-containers && \
sudo chmod 0755 /opt/kata /opt/kata/bin /opt/kata/share /opt/kata/share/defaults /opt/kata/share/defaults/kata-containers && \
sudo cp target/debug/genpolicy /opt/kata/bin/ && \
sudo chmod 0755 /opt/kata/bin/genpolicy && \
sudo cp ./rules.rego /opt/kata/share/defaults/kata-containers/ && \
sudo chmod 0644 /opt/kata/share/defaults/kata-containers/rules.rego && \
sudo cp ./genpolicy-settings.json /opt/kata/share/defaults/kata-containers/ && \
sudo chmod 0644 /opt/kata/share/defaults/kata-containers/genpolicy-settings.json

popd
```

Execute tests:

```
sudo tdnf install -y jq

git clone https://github.com/bats-core/bats-core.git
cd bats-core
sudo ./install.sh /usr/local

cd kata-containers/tests/integration/kubernetes/

AUTO_GENERATE_POLICY=yes \
AZ_RG="${my_resource_group}" \
DOCKER_TAG=${DOCKER_TAG} \
GH_PR_NUMBER=${GH_PR_NUMBER} \
DOCKER_REGISTRY=ghcr.io \
DOCKER_REPO=kata-containers/kata-deploy-ci \
bash gha-run.sh run-tests
```

If you just want to execute k8s-policy-job.bats, prepend K8S_TEST_UNION="k8s-policy-job.bats" to the run-tests command line above.

**Comment by burgerdev** [comment]
Ok, I still need to figure out what the actually expected agent bundle paths are. [This test](https://github.com/kata-containers/kata-containers/actions/runs/10278843023/job/28499762264?pr=10050#step:9:1963) tries to add files to `/run/kata-containers/{blob}` but the regex expects it at `/run/kata-containers/shared/containers/{blob}` (which would have been my expectation, too).

```
#   Warning  FailedCreatePodSandBox  81s   kubelet            Failed to create pod sandbox: rpc error: code = Unknown desc = failed to create containerd task: failed to create shim task: Could not send CopyFile request: rpc error: code = PermissionDenied desc = "CopyFileRequest is blocked by policy: agent_policy:1101:  CopyFileRequest: input.path = /run/kata-containers/shared/containers/57168708f0677f3f7a1758fe173fd48813d8210a31f1b6ebd132eb723fc3f453-206077e202f30e38-resolv.conf agent_policy:1109:  CopyFileRequest: regex4 = ^/run/kata-containers/[a-z0-9]{64}-[a-z0-9]{16}-": unknown
```

**Comment by burgerdev** [comment]
I think I understood now: `cpath` is the place where the container rootfs go, while `mount_source_cpath` is where files are copied (to then be bind-mounted into the container). We might want to standardize the string interpolation in `genpolicy-settings.json` eventually, as its not clear from looking at the settings file which of the two values is going to replace `$(cpath)`.

File: src/tools/genpolicy/rules.rego [hunk 1]
```diff
@@ -1092,12 +1092,23 @@ match_caps(p_caps, i_caps) {
 }
 
 ######################################################################
+
+check_directory_traversal(i_path) {
+    not regex.match("(^|/)..($|/)", i_path)
```
Comment (danmihai1) [inline src/tools/genpolicy/rules.rego:1097]: @burgerdev , are there other advantages to using this regex, except for a shorter policy text?

If there aren't other advantages, this other version is easier for me to understand:

check_directory_traversal(i_path) {
    contains(i_path, "../") == false
    endswith(i_path, "/..") == false
    i_path != ".."
}

Do you feel strongly that we should use the regex?

File: src/tools/genpolicy/tests/testdata/copyfile/testcases.json [hunk 2]
```diff
@@ -12,5 +12,26 @@
...
+    "request": {
+      "path": "/run/kata-containers/81e5f43bc8599c5661e66f959ac28df5bfb30da23c5d583f2dcc6f9e0c5186dc/rootfs/bin/sh"
+    }
+  },
+  {
+    "description": "attempted directory traversal",
+    "allowed": false,
+    "request": {
+      "path": "/run/kata-containers/shared/containers/81e5f43bc8599c5661e66f959ac28df5bfb30da23c5d583f2dcc6f9e0c5186dc-ce23cfeb91e75aaa-foo/../../../../../etc/ssl/cert.pem"
+    }
```
Comment (danmihai1) [inline src/tools/genpolicy/tests/testdata/copyfile/testcases.json:35]: Please add two more test cases for:

endswith(i_path, "/..") == false
i_path != ".."

File: tests/integration/kubernetes/tests_common.sh [hunk 3]
```diff
@@ -97,7 +97,12 @@ delete_pod_if_new() {
 exec_host() {
 	node="$1"
 	# `kubectl debug` always returns 0, so we hack it to return the right exit code.
-	command="${@:2}"
+	command="${*:2}"
```
Comment (danmihai1) [inline tests/integration/kubernetes/tests_common.sh:100]: If those changes are not directly related to the Policy changes, I would prefer to move them into a separate PR. @GabyCT has been trying to improve this function too, related to https://github.com/kata-containers/kata-containers/issues/10081.

File: src/tools/genpolicy/rules.rego (hunk 1 above)
Comment (burgerdev) [reply src/tools/genpolicy/rules.rego:1097]: I wrote the regex when I realized the above is wrong: consider `i_path == /foo/bar/etc../file`. Let me add that to the test cases.

File: tests/integration/kubernetes/tests_common.sh (hunk 3 above)
Comment (burgerdev) [reply tests/integration/kubernetes/tests_common.sh:100]: Yeah, I was just annoyed by these non-descriptive CI failures and wanted to see if I can do something about it. Happy to remove.

**Comment by burgerdev** [comment]
Looking again at the CI runs, I see

* sporadic errors like reported in #10081 
* Nydus errors for absent image decryption keys: https://github.com/kata-containers/kata-containers/actions/runs/10365121178/job/28692498816?pr=10050#step:9:2351
* CoCo test failures that lack debug information: https://github.com/kata-containers/kata-containers/actions/runs/10365121178/job/28692498816?pr=10050#step:9:11069

@danmihai1, wdyt?

**Review by sprt — APPROVED** [review]
